diff --git a/operators/xr/python/xr_begin_frame_pydoc.hpp b/operators/xr/python/xr_begin_frame_pydoc.hpp
index 20d1d4ca5..65e6dc63e 100644
--- a/operators/xr/python/xr_begin_frame_pydoc.hpp
+++ b/operators/xr/python/xr_begin_frame_pydoc.hpp
@@ -34,8 +34,8 @@ XrSession obj representing the current state of the xr system
 
 Parameters
 ----------
-fragment : Fragment
-The fragment that the operator belongs to.
+fragment_or_subgraph : Fragment or Subgraph
+The fragment or subgraph that the operator belongs to.
 xr_session : ``holoscan.XrSession``
 The shared XrSession object
 name : str, optional
diff --git a/operators/xr/python/xr_controller_tracker_pydoc.hpp b/operators/xr/python/xr_controller_tracker_pydoc.hpp
new file mode 100644
index 000000000..17e31a2d4
--- /dev/null
+++ b/operators/xr/python/xr_controller_tracker_pydoc.hpp
@@ -0,0 +1,39 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+namespace holoscan::doc::XrControllerTracker {
+
+constexpr const char* doc_XrControllerTracker_python = R"doc(
+Initialize an XR controller tracker resource.
+
+Parameters
+----------
+fragment_or_subgraph : Fragment or Subgraph
+    The fragment or subgraph that the resource belongs to.
+xr_session : holoscan.XrSession
+    The XR session resource.
+hand : XrHandEXT
+    Which hand to track.
+reference_space : XrReferenceSpaceType, optional
+    Reference space to use for tracking. Default is LOCAL.
+name : str, optional
+    The name of the resource.
+)doc";
+
+}  // namespace holoscan::doc::XrControllerTracker
diff --git a/operators/xr/python/xr_empty_composition_layer_pydoc.hpp b/operators/xr/python/xr_empty_composition_layer_pydoc.hpp
index 237cac3cd..33eacbd52 100644
--- a/operators/xr/python/xr_empty_composition_layer_pydoc.hpp
+++ b/operators/xr/python/xr_empty_composition_layer_pydoc.hpp
@@ -36,8 +36,8 @@ layer that can be processed by XrEndFrameOp.
 
 Parameters
 ----------
-fragment : Fragment
-    The fragment that the operator belongs to.
+fragment_or_subgraph : Fragment or Subgraph
+    The fragment or subgraph that the operator belongs to.
 name : str, optional
     The name of the operator. Default value is ``"xr_empty_composition_layer_op"``.
 )doc";
diff --git a/operators/xr/python/xr_end_frame_pydoc.hpp b/operators/xr/python/xr_end_frame_pydoc.hpp
index bb3798652..f98607666 100644
--- a/operators/xr/python/xr_end_frame_pydoc.hpp
+++ b/operators/xr/python/xr_end_frame_pydoc.hpp
@@ -34,8 +34,8 @@ XrSession obj representing the current state of the xr system
 
 Parameters
 ----------
-fragment : Fragment
-The fragment that the operator belongs to.
+fragment_or_subgraph : Fragment or Subgraph
+The fragment or subgraph that the operator belongs to.
 xr_session : ``holoscan.XrSession``
 The shared XrSession object
 xr_composition_layers : ``IOSpec*``
diff --git a/operators/xr/python/xr_hand_tracker_pydoc.hpp b/operators/xr/python/xr_hand_tracker_pydoc.hpp
new file mode 100644
index 000000000..32ea8ac66
--- /dev/null
+++ b/operators/xr/python/xr_hand_tracker_pydoc.hpp
@@ -0,0 +1,39 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+namespace holoscan::doc::XrHandTracker {
+
+constexpr const char* doc_XrHandTracker_python = R"doc(
+Initialize an XR hand tracker resource.
+
+Parameters
+----------
+fragment_or_subgraph : Fragment or Subgraph
+    The fragment or subgraph that the resource belongs to.
+xr_session : holoscan.XrSession
+    The XR session resource.
+hand : XrHandEXT
+    Which hand to track.
+reference_space : XrReferenceSpaceType, optional
+    Reference space to use for tracking. Default is LOCAL.
+name : str, optional
+    The name of the resource.
+)doc";
+
+}  // namespace holoscan::doc::XrHandTracker
diff --git a/operators/xr/python/xr_python.cpp b/operators/xr/python/xr_python.cpp
index ab8eb59a9..0477ebe62 100644
--- a/operators/xr/python/xr_python.cpp
+++ b/operators/xr/python/xr_python.cpp
@@ -21,23 +21,28 @@
 
 #include <memory>
 #include <string>
+#include <variant>
 #include <vector>
 
 #include <xr_begin_frame_op.hpp>
 #include <xr_begin_frame_pydoc.hpp>
 #include <xr_composition_layers.hpp>
 #include <xr_controller_tracker.hpp>
+#include <xr_controller_tracker_pydoc.hpp>
 #include <xr_empty_composition_layer_op.hpp>
 #include <xr_empty_composition_layer_pydoc.hpp>
 #include <xr_end_frame_op.hpp>
 #include <xr_end_frame_pydoc.hpp>
 #include <xr_hand_tracker.hpp>
+#include <xr_hand_tracker_pydoc.hpp>
 #include <xr_swapchain_cuda.hpp>
 #include <../../operator_util.hpp>
 
 #include <holoscan/core/fragment.hpp>
+#include <holoscan/core/subgraph.hpp>
 #include <holoscan/core/operator.hpp>
 #include <holoscan/core/operator_spec.hpp>
+#include <holoscan/python/core/component_util.hpp>
 #include <holoscan/python/core/emitter_receiver_registry.hpp>
 
 namespace py = pybind11;
@@ -51,15 +56,13 @@ class PyXrBeginFrameOp : public XrBeginFrameOp {
   using XrBeginFrameOp::XrBeginFrameOp;
 
   // Define a constructor that fully initializes the object.
-  PyXrBeginFrameOp(Fragment* fragment, const py::args& args,
+  PyXrBeginFrameOp(const std::variant<holoscan::Fragment*, holoscan::Subgraph*>& fragment_or_subgraph,
+                   const py::args& args,
                    std::shared_ptr<holoscan::XrSession> xr_session,
                    const std::string& name = "xr_begin_frame")
       : XrBeginFrameOp(ArgList{Arg{"xr_session", xr_session}}) {
     add_positional_condition_and_resource_args(this, args);
-    name_ = name;
-    fragment_ = fragment;
-    spec_ = std::make_shared<OperatorSpec>(fragment);
-    setup(*spec_.get());
+    holoscan::init_operator_base(this, fragment_or_subgraph, name);
   }
 };
 
@@ -69,15 +72,13 @@ class PyXrEndFrameOp : public XrEndFrameOp {
   using XrEndFrameOp::XrEndFrameOp;
 
   // Define a constructor that fully initializes the object.
-  PyXrEndFrameOp(Fragment* fragment, const py::args& args,
+  PyXrEndFrameOp(const std::variant<holoscan::Fragment*, holoscan::Subgraph*>& fragment_or_subgraph,
+                 const py::args& args,
                  std::shared_ptr<holoscan::XrSession> xr_session,
                  const std::string& name = "xr_end_frame")
       : XrEndFrameOp(ArgList{Arg{"xr_session", xr_session}}) {
     add_positional_condition_and_resource_args(this, args);
-    name_ = name;
-    fragment_ = fragment;
-    spec_ = std::make_shared<OperatorSpec>(fragment);
-    setup(*spec_.get());
+    holoscan::init_operator_base(this, fragment_or_subgraph, name);
   }
 };
 
@@ -87,14 +88,13 @@ class PyXrEmptyCompositionLayerOp : public XrEmptyCompositionLayerOp {
   using XrEmptyCompositionLayerOp::XrEmptyCompositionLayerOp;
 
   // Define a constructor that fully initializes the object.
-  PyXrEmptyCompositionLayerOp(Fragment* fragment, const py::args& args,
-                               const std::string& name = "xr_empty_composition_layer")
+  PyXrEmptyCompositionLayerOp(
+                              const std::variant<holoscan::Fragment*, holoscan::Subgraph*>& fragment_or_subgraph,
+                              const py::args& args,
+                              const std::string& name = "xr_empty_composition_layer")
       : XrEmptyCompositionLayerOp() {
     add_positional_condition_and_resource_args(this, args);
-    name_ = name;
-    fragment_ = fragment;
-    spec_ = std::make_shared<OperatorSpec>(fragment);
-    setup(*spec_.get());
+    holoscan::init_operator_base(this, fragment_or_subgraph, name);
   }
 };
 
@@ -106,16 +106,15 @@ class PyXrSession : public XrSession {
  public:
   using XrSession::XrSession;
 
-  PyXrSession(Fragment* fragment, const std::string& application_name = "Holoscan XR",
-              uint32_t application_version = 0u, const std::string& name = "xr_session")
+  PyXrSession(const std::variant<holoscan::Fragment*, holoscan::Subgraph*>& fragment_or_subgraph,
+              const std::string& application_name = "Holoscan XR",
+              uint32_t application_version = 0u,
+              const std::string& name = "xr_session")
       : XrSession(ArgList{
             Arg{"application_name", application_name},
             Arg{"application_version", application_version},
         }) {
-    name_ = name;
-    fragment_ = fragment;
-    spec_ = std::make_shared<ComponentSpec>(fragment);
-    setup(*spec_.get());
+    holoscan::init_component_base(this, fragment_or_subgraph, name, "resource");
   }
 };
 
@@ -123,14 +122,13 @@ class PyXrHandTracker : public XrHandTracker {
  public:
   using XrHandTracker::XrHandTracker;
 
-  PyXrHandTracker(Fragment* fragment, std::shared_ptr<holoscan::XrSession> xr_session,
-                  xr::HandEXT hand, xr::ReferenceSpaceType reference_space = xr::ReferenceSpaceType::Local,
-                  const std::string& name = "xr_hand_tarcker")
+  PyXrHandTracker(const std::variant<holoscan::Fragment*, holoscan::Subgraph*>& fragment_or_subgraph,
+                  std::shared_ptr<holoscan::XrSession> xr_session,
+                  xr::HandEXT hand,
+                  xr::ReferenceSpaceType reference_space = xr::ReferenceSpaceType::Local,
+                  const std::string& name = "xr_hand_tracker")
       : XrHandTracker(xr_session, hand, reference_space) {
-    name_ = name;
-    fragment_ = fragment;
-    spec_ = std::make_shared<ComponentSpec>(fragment);
-    setup(*spec_.get());
+    holoscan::init_component_base(this, fragment_or_subgraph, name, "resource");
   }
 };
 
@@ -138,14 +136,14 @@ class PyXrControllerTracker : public XrControllerTracker {
  public:
   using XrControllerTracker::XrControllerTracker;
 
-  PyXrControllerTracker(Fragment* fragment, std::shared_ptr<holoscan::XrSession> xr_session,
-                        xr::HandEXT hand, xr::ReferenceSpaceType reference_space = xr::ReferenceSpaceType::Local,
-                        const std::string& name = "xr_controller_tracker")
+  PyXrControllerTracker(
+      const std::variant<holoscan::Fragment*, holoscan::Subgraph*>& fragment_or_subgraph,
+      std::shared_ptr<holoscan::XrSession> xr_session,
+      xr::HandEXT hand,
+      xr::ReferenceSpaceType reference_space = xr::ReferenceSpaceType::Local,
+      const std::string& name = "xr_controller_tracker")
       : XrControllerTracker(xr_session, hand, reference_space) {
-    name_ = name;
-    fragment_ = fragment;
-    spec_ = std::make_shared<ComponentSpec>(fragment);
-    setup(*spec_.get());
+    holoscan::init_component_base(this, fragment_or_subgraph, name, "resource");
   }
 };
 
@@ -209,7 +207,7 @@ PYBIND11_MODULE(_xr, m) {
              Operator,
              std::shared_ptr<ops::XrBeginFrameOp>>(
       m, "XrBeginFrameOp", doc::XrBeginFrameOp::doc_XrBeginFrameOp_python)
-      .def(py::init<Fragment*,
+      .def(py::init<const std::variant<holoscan::Fragment*, holoscan::Subgraph*>&,
                     const py::args&,
                     std::shared_ptr<holoscan::XrSession>,
                     const std::string&>(),
@@ -219,7 +217,7 @@ PYBIND11_MODULE(_xr, m) {
            doc::XrBeginFrameOp::doc_XrBeginFrameOp_python);
   py::class_<ops::XrEndFrameOp, ops::PyXrEndFrameOp, Operator, std::shared_ptr<ops::XrEndFrameOp>>(
       m, "XrEndFrameOp", doc::XrEndFrameOp::doc_XrEndFrameOp_python)
-      .def(py::init<Fragment*,
+      .def(py::init<const std::variant<holoscan::Fragment*, holoscan::Subgraph*>&,
                     const py::args&,
                     std::shared_ptr<holoscan::XrSession>,
                     const std::string&>(),
@@ -235,13 +233,18 @@ PYBIND11_MODULE(_xr, m) {
       "XrEmptyCompositionLayerOp",
       doc::XrEmptyCompositionLayerOp::doc_XrEmptyCompositionLayerOp_python)
       .def(
-          py::init<Fragment*, const py::args&, const std::string&>(),
+          py::init<const std::variant<holoscan::Fragment*, holoscan::Subgraph*>&,
+                   const py::args&,
+                   const std::string&>(),
           "fragment"_a,
           "name"_a = "xr_empty_composition_layer_op"s,
           doc::XrEmptyCompositionLayerOp::doc_XrEmptyCompositionLayerOp_python);
 
   py::class_<XrSession, PyXrSession, holoscan::Resource, std::shared_ptr<XrSession>>(m, "XrSession")
-      .def(py::init<Fragment*, const std::string&, uint32_t, const std::string&>(),
+      .def(py::init<const std::variant<holoscan::Fragment*, holoscan::Subgraph*>&,
+                    const std::string&,
+                    uint32_t,
+                    const std::string&>(),
            "fragment"_a,
            "application_name"_a = "Holoscan XR",
            "application_version"_a = 0u,
@@ -378,7 +381,7 @@ PYBIND11_MODULE(_xr, m) {
 
   py::class_<XrHandTracker, PyXrHandTracker, holoscan::Resource, std::shared_ptr<XrHandTracker>>(
       m, "XrHandTracker")
-      .def(py::init<Fragment*,
+      .def(py::init<const std::variant<holoscan::Fragment*, holoscan::Subgraph*>&,
                     std::shared_ptr<holoscan::XrSession>,
                     xr::HandEXT,
                     xr::ReferenceSpaceType,
@@ -387,7 +390,8 @@ PYBIND11_MODULE(_xr, m) {
            "xr_session"_a,
            "hand"_a,
            "reference_space"_a = xr::ReferenceSpaceType::Local,
-           "name"_a = "xr_hand_tracker"s)
+           "name"_a = "xr_hand_tracker"s,
+           doc::XrHandTracker::doc_XrHandTracker_python)
       .def("locate_hand_joints", &XrHandTracker::locate_hand_joints);
 
   py::class_<ControllerState>(m, "ControllerState")
@@ -418,7 +422,7 @@ PYBIND11_MODULE(_xr, m) {
              PyXrControllerTracker,
              holoscan::Resource,
              std::shared_ptr<XrControllerTracker>>(m, "XrControllerTracker")
-      .def(py::init<Fragment*,
+      .def(py::init<const std::variant<holoscan::Fragment*, holoscan::Subgraph*>&,
                     std::shared_ptr<holoscan::XrSession>,
                     xr::HandEXT,
                     xr::ReferenceSpaceType,
@@ -427,7 +431,8 @@ PYBIND11_MODULE(_xr, m) {
            "xr_session"_a,
            "hand"_a,
            "reference_space"_a = xr::ReferenceSpaceType::Local,
-           "name"_a = "xr_controller_tracker"s)
+           "name"_a = "xr_controller_tracker"s,
+           doc::XrControllerTracker::doc_XrControllerTracker_python)
       .def("get_controller_state", &XrControllerTracker::get_controller_state);
 
   py::enum_<XrSwapchainCuda::Format>(m, "XrSwapchainCudaFormat")
